File: prototype/krylov_block.py
#!/usr/bin/env python3
"""Clean block krylov solver implementation.

This can replace krylov_glm in the CPHF notebook. It solves (1+a)*x = b
using a block Krylov approach that adds nroots directions per cycle,
matching PySCF's lib.krylov convergence behavior.

Key differences from krylov_glm (GMRES):
1. Block approach: adds all RHS directions per cycle, not one per iteration
2. Non-normalized basis: trial vectors deflate naturally as solution converges
3. Identity handled analytically: Krylov subspace built for `a` only

Usage:
    # Instead of:
    mo1 = krylov_glm(vind_vo_plus, mo1_base, tol=1e-8)

    # Use:
    mo1 = krylov_block(vind_vo, mo1_base, tol=1e-8)

    # Note: pass vind_vo (without +I), not vind_vo_plus!
"""

import numpy as np


def krylov_block(aop, b, x0=None, tol=1e-10, max_cycle=30, lindep=1e-13):
    """Block Krylov subspace method to solve (1+a)*x = b.

    At each cycle, the current trial block is multiplied by aop and
    orthogonalized against the full subspace.  This adds nroots new
    directions per cycle (one per surviving right-hand side), making
    it much faster than single-vector GMRES for multi-RHS problems.

    The basis vectors are kept non-normalized (orthogonal but not
    orthonormal), and their squared norms are tracked in `innerprod`.
    This causes trial vectors to naturally "deflate" as the solution
    converges, providing an accurate and inexpensive convergence signal
    without requiring the actual residual to be computed.

    Parameters
    ----------
    aop : callable
        Linear operator a, mapping (nblock, n) -> (nblock, n).
        The equation solved is (1 + aop) * x = b.
    b : ndarray, shape (nset, n) or (n,)
        Right-hand sides.
    x0 : ndarray, shape (nset, n) or (n,), optional
        Initial guess.
    tol : float
        Convergence tolerance.  Iteration stops when
        max(||new_trial_vec_i||^2) < max(lindep, tol^2).
    max_cycle : int
        Maximum number of block cycles.
    lindep : float
        Linear dependency threshold.  Vectors with ||v||^2 < lindep
        are dropped from the subspace.

    Returns
    -------
    x : ndarray, same shape as b
        Approximate solution of (1 + aop) * x = b.
    """
    if b.ndim == 1:
        b = b.reshape(1, -1)
        was_1d = True
    else:
        was_1d = False
    nset_l, ndim_l = b.shape

    if x0 is not None:
        if x0.ndim == 1:
            x0 = x0.reshape(1, -1)
        b = b - (x0 + aop(x0))

    # MGS that keeps non-normalized vectors and tracks ||v||^2
    def _orth_block(vec_list):
        result = []
        norms_sq = []
        for vi in vec_list:
            vi = vi.copy()
            for j in range(len(result)):
                coeff = np.dot(vi, result[j]) / norms_sq[j]
                vi -= coeff * result[j]
            nsq = np.dot(vi, vi).real
            if nsq > lindep:
                result.append(vi)
                norms_sq.append(nsq)
        return result, norms_sq

    # Initialize: orthogonalize b
    x1_list, innerprod = _orth_block([b[i] for i in range(nset_l)])

    if not x1_list:
        result = np.zeros((nset_l, ndim_l))
        if x0 is not None:
            result += x0
        return result[0] if was_1d else result

    xs = []              # basis vectors (orthogonal, non-normalized)
    ax_list = []         # aop(xs[i]) for each basis vector
    all_innerprod = []    # ||xs[i]||^2

    for cycle in range(max_cycle):
        # Apply operator to current trial block
        x1_arr = np.array(x1_list)
        axt = aop(x1_arr)
        if axt.ndim == 1:
            axt = axt.reshape(1, -1)

        # Store basis vectors and aop results
        for i in range(len(x1_list)):
            xs.append(x1_list[i].copy())
            ax_list.append(axt[i].copy())
            all_innerprod.append(innerprod[i])

        # Orthogonalize axt against full subspace
        # For non-normalized orthogonal xs[i] with ||xs[i]||^2 = all_innerprod[i]:
        #   proj_coeff = dot(v, xs[i]) / all_innerprod[i]
        x1_new = axt.copy()
        for i in range(len(xs)):
            xsi = xs[i]
            w = x1_new @ xsi / all_innerprod[i]    # (nblock,) coefficients
            x1_new -= np.outer(w, xsi)

        # Orthogonalize among themselves via MGS
        x1_list, innerprod = _orth_block(x1_new)

        max_innerprod = max(innerprod) if innerprod else 0
        r = np.sqrt(max_innerprod)

        print(f"Cycle {cycle+1}: max(||new_trial_vec_i||^2) = {max_innerprod:.3e}, max(||new_trial_vec_i||) = {r:.3e}")

        # Convergence check (same as PySCF)
        if max_innerprod < max(lindep, tol**2):
            break

    # Build and solve small system: (I + A_projected) * c = b_projected
    nd = len(xs)
    Xs = np.array(xs)
    AX = np.array(ax_list)

    # h[i,j] = dot(xs[i], aop(xs[j])) + delta_ij * ||xs[i]||^2
    h = Xs @ AX.T
    for i in range(nd):
        h[i, i] += all_innerprod[i]

    # g[i,k] = dot(b[k], xs[i])
    g = Xs @ b.T

    c = np.linalg.solve(h, g)
    x = c.T @ Xs

    if x0 is not None:
        x += x0

    return x[0] if was_1d else x

File: src/util/krylov_block.rs
//! Block Krylov subspace solver for `(1 + A) x = b`.
//!
//! Direct translation of `prototype/krylov_block.py`. The convergence behavior
//! matches PySCF's `lib.krylov` block algorithm: each cycle adds the surviving
//! trial directions to the subspace, and the basis is kept non-normalized so
//! that the squared norms of the new trial vectors act as the convergence
//! signal without requiring an explicit residual evaluation.
//!
//! Layout convention is col-major: each right-hand side / basis vector is a
//! **column** of the corresponding matrix. So `b` is shaped `[n, nset]`,
//! the operator maps `[n, nblock] -> [n, nblock]`, and the basis matrices
//! `xs`, `ax` are `[n, nd]` with new vectors appended along axis 1.

use crate::prelude::*;

/// Solve `(I + aop) x = b` by a block Krylov subspace method.
///
/// # Parameters
///
/// - `aop` : Linear operator. Given a `[n, nblock]` input it must return a `[n, nblock]` output
///   (the action of `A` applied column-wise).
/// - `b` : Right-hand sides, shape `[n, nset]`. Each column is one RHS.
/// - `x0` : Optional initial guess, shape `[n, nset]`.
/// - `tol` : Convergence tolerance on `max(||new_trial_vec_i||)`.
/// - `max_cycle` : Maximum number of block cycles.
/// - `lindep` : Vectors with `||v||^2 < lindep` are dropped from the subspace.
///
/// # Returns
///
/// `x` of shape `[n, nset]`, an approximate solution of `(I + aop) x = b`.
pub fn krylov_block(
    mut aop: impl FnMut(TsrView) -> Tsr,
    b: TsrView,
    x0: Option<TsrView>,
    tol: f64,
    max_cycle: usize,
    lindep: f64,
) -> Tsr {
    let device = b.device().clone();
    let ndim_l = b.shape()[0];
    let nset_l = b.shape()[1];

    // Subtract the contribution of the initial guess from the RHS.
    let b: Tsr = match x0.as_ref() {
        Some(x0v) => &b - (&x0v.to_owned() + aop(x0v.view())),
        None => b.to_owned(),
    };

    // Initialize: orthogonalize the columns of b.
    let (mut x1, mut innerprod) = orth_block(b.view(), lindep);

    if x1.shape()[1] == 0 {
        let mut result: Tsr = rt::zeros(([ndim_l, nset_l], &device));
        if let Some(x0v) = x0 {
            result += x0v;
        }
        return result;
    }

    // Pre-allocate basis storage: at most nset_l vectors per cycle plus the
    // initial block, so cap at nset_l * (max_cycle + 1) columns.
    let max_basis = nset_l * (max_cycle + 1);
    let mut xs: Tsr = rt::zeros(([ndim_l, max_basis], &device));
    let mut ax: Tsr = rt::zeros(([ndim_l, max_basis], &device));
    let mut all_innerprod: Vec<f64> = Vec::with_capacity(max_basis);
    let mut nd: usize = 0;

    let conv_thresh = lindep.max(tol * tol);

    for cycle in 0..max_cycle {
        let nblock = x1.shape()[1];

        // Apply operator to current trial block.
        let axt = aop(x1.view());

        // Append current (x1, axt, innerprod) to the subspace.
        xs.i_mut((.., nd..nd + nblock)).assign(&x1);
        ax.i_mut((.., nd..nd + nblock)).assign(&axt);
        all_innerprod.extend_from_slice(&innerprod);
        nd += nblock;

        // Orthogonalize axt against the full subspace. For non-normalized
        // orthogonal columns xs[:, i] with ||xs[:, i]||^2 = all_innerprod[i]:
        //   coeffs[i, k] = (xs[:, i] . axt[:, k]) / all_innerprod[i]
        //              = (xs_slc.T @ axt)[i, k] / all_innerprod[i]
        //   x1_new[:, k] = axt[:, k] - sum_i coeffs[i, k] * xs[:, i]
        //              = axt - xs_slc @ coeffs
        let xs_slc = xs.i((.., ..nd));
        let ip_vec: Tsr = rt::asarray((all_innerprod.clone(), &device));
        let coeffs = (xs_slc.t() % &axt) / ip_vec.i((.., None));
        let x1_new = axt - &xs_slc % &coeffs;

        // Orthogonalize the new trial block among itself.
        let (next_x1, next_ip) = orth_block(x1_new.view(), lindep);

        let max_innerprod = next_ip.iter().copied().fold(0.0_f64, f64::max);
        let r = max_innerprod.sqrt();
        println!(
            "Cycle {}: max(||new_trial_vec_i||^2) = {:.3e}, max(||new_trial_vec_i||) = {:.3e}",
            cycle + 1,
            max_innerprod,
            r
        );

        x1 = next_x1;
        innerprod = next_ip;

        if max_innerprod < conv_thresh {
            break;
        }
    }

    // Build and solve the projected system: (I + A_projected) c = b_projected.
    let xs_slc = xs.i((.., ..nd));
    let ax_slc = ax.i((.., ..nd));

    // h[i, j] = dot(xs[:, i], ax[:, j]) + delta_ij * ||xs[:, i]||^2
    //        = (xs_slc.T @ ax_slc)[i, j] + diagonal correction
    let mut h: Tsr = xs_slc.t() % &ax_slc;
    for i in 0..nd {
        h[[i, i]] += all_innerprod[i];
    }

    // g[i, k] = dot(b[:, k], xs[:, i]) = (xs_slc.T @ b)[i, k]
    let g: Tsr = xs_slc.t() % &b;

    // Solve h c = g, then reconstruct x[:, k] = sum_i c[i, k] * xs[:, i] = xs c.
    let c = rt::linalg::solve_general((h.view(), g.view()));
    let mut x: Tsr = &xs_slc % &c;

    if let Some(x0v) = x0 {
        x += x0v;
    }
    x
}

/// Modified Gram-Schmidt over the **columns** of `vec`, keeping non-normalized
/// orthogonal vectors and tracking their squared norms. Mirrors `_orth_block`
/// in the Python prototype (which operated on rows; here we operate on columns
/// because the rest of the solver is col-major).
///
/// Columns whose remaining squared norm falls below `lindep` are dropped.
///
/// Returns `(out, norms_sq)` where `out` has shape `[n, m]` with `m <= nblock`.
fn orth_block(vec: TsrView, lindep: f64) -> (Tsr, Vec<f64>) {
    let device = vec.device().clone();
    let n = vec.shape()[0];
    let nblock = vec.shape()[1];

    let mut result: Vec<Tsr> = Vec::with_capacity(nblock);
    let mut norms_sq: Vec<f64> = Vec::with_capacity(nblock);

    for i in 0..nblock {
        let mut vi: Tsr = vec.i((.., i)).to_owned();
        for j in 0..result.len() {
            let coeff = (&vi % &result[j]).to_scalar() / norms_sq[j];
            vi -= coeff * &result[j];
        }
        let nsq = (&vi % &vi).to_scalar();
        if nsq > lindep {
            result.push(vi);
            norms_sq.push(nsq);
        }
    }

    let m = result.len();
    let mut out: Tsr = rt::zeros(([n, m], &device));
    for (i, vi) in result.into_iter().enumerate() {
        out.i_mut((.., i)).assign(&vi);
    }
    (out, norms_sq)
}
